### PR TITLE
[msh3,msquic,vcpkg-ci-msh3] Update, add test port

### LINKED ISSUE #41200: [msh3] update to 0.7.0
### Library name

msh3

### New version number

0.7.0

### Other information that may be useful (release notes, etc...)

it needed to change also msquick:
https://github.com/microsoft/msquic/blob/v2.3.5/src/inc/quic_platform.h#L574
That needed c++ define:
https://github.com/microsoft/msquic/blob/v2.3.5/src/inc/quic_platform.h#L498

Not sure how to do that.

When I tried to update:
```
msh3.cpp.obj : error LNK2001: unresolved external symbol CxPlatGetSelfSignedCert
msh3.cpp.obj : error LNK2001: unresolved external symbol CxPlatFreeSelfSignedCert
bin\msh3.dll : fatal error LNK1120: 2 unresolved externals

ninja: build stopped: subcommand failed.
```

I am not sure how to compile these symbol. I know they are depends on define cpp: QUIC_TEST_APIS
https://github.com/microsoft/msquic/blob/v2.3.5/src/inc/quic_platform.h#L498

I tried to add it via port config with `VCPKG_CXX_FLAGS` with both define in msquic and msh3, but without success.

### LINKED ISSUE #42755: [msquic] update to 2.4.7
### Library name

msquic

### New version number

2.4.7

### Other information that may be useful (release notes, etc...)

My attempt to do this:
https://github.com/microsoft/vcpkg/pull/42547
You can use the code and continue my work.

### PR DESCRIPTION
**Pull request by dg0yt**
- [x] Establish test port, fix current version of msquic/msh3.
- [x] Update msquic. Resolves #42755.
- [x] Update msh3. Resolves #41200.
- [x] Enable uwp.

### CODE REVIEW

**Comment by dg0yt** [comment]
@nibanks This is the sad state of port msh3's exported config in vcpkg.

(I can push fixes and updates once CI completes. This is my only source of MSVC logs.)

**Comment by nibanks** [comment]
> @nibanks This is the sad state of port msh3's exported config in vcpkg.
> 
> (I can push fixes and updates once CI completes. This is my only source of MSVC logs.)

What is needed from msh3 to make this better?

**Comment by dg0yt** [comment]
> What is needed from msh3 to make this better?

The only immediate suggestuion is to have a similar test upstream: Add a CI step which configures and builds (and maybe runs) an independent CMake test project based on the canonical CMake usage pattern. This catches missing properites and transitive usage requirements. (If transitive linking requirements are the blocker for offering a static lib, *this* would be a building block to improve the situation.) Similar for pkg-config.

This test port passes now, and I can update msh3 here on that base. Let's see what is still needed then. Of course it would be nice to not need patching in order to de-vendor msquic and ls-qpack, and to simply follow explicit toolchain choices.

**Comment by dg0yt** [comment]
Oh, great. Upgrade msh3 needs upgrading msquic. And we get a wonderful clash of the non-namespaced CMake targets. Needs more work.

**Comment by dg0yt** [comment]
After updating msquic (WIP):
~~~
msquic provides CMake targets:

  # this is heuristically generated, and may not be correct
  find_package(msquic CONFIG REQUIRED)
  # note: 4 additional targets are not displayed.
  target_link_libraries(main PRIVATE inc msquic OpenSSL warnings)
~~~
I would assume that `msquic` should be the public CMake interface, pulling in everything which is needed to use msquic.

The port only support dynamic library linkage ATM. Upstream supports both, but doesn't prvovide exported config for the static lib. It is not create with `add_library`. In general, it would be desirable to have the same target name, with proper namespace, i.e. `msquic::msquic`.

**Comment by dg0yt** [comment]
#41219 disabled uwp for msquic due to (a variation of) this error:
~~~
OneCore.lib(KERNEL32.dll) : error LNK2005: VirtualAlloc already defined in msquic_platform.lib(datapath_winuser.c.obj)
   Creating library obj\Debug\msquic.lib and object obj\Debug\msquic.exp
bin\Debug\msquic.dll : fatal error LNK1169: one or more multiply defined symbols found
~~~
The error didn't really make sense to me wrt msquic_platform.lib(datapath_winuser.c.obj), and I started to look at other link libs appearing after msquic_platform.lib until (incl.) OneCore.lib. I also looked at GHA logs in the msquic repo. (But they don't use the ninja generator.)
I was now able to track this down to the use of "ONECORE" instead of "UWP" in openssl3.

@nibanks Maybe you can help answer questions also for msquic? I'm always a litte bit last in the past, present and future of Microsoft platforms.
- Choose schannel or openssl as default windows configuruation?
  Should the default installation of the msquic port on Windows use openssl3?
  AFAIU msquic default choice is schannel, but this choice doesn't offer the 0-RTT capability,  
  https://github.com/microsoft/msquic/blob/main/docs/Platforms.md  
  (vcpkg CI could still run with the capability enabled, via the test port requesting a new `0-rtt` feature.)
- Link OneCore, OneCoreUap, or WindowsApp?
  AFAIU OneCore must not be chosen for UWP,
  https://learn.microsoft.com/en-us/windows/win32/apiindex/windows-umbrella-libraries  
  The linker error shown above disappers when configuring openssl for UWP instead of ONECORE. (This is what the pristine openssl vcpkg port does.)
- Avoid winmm.lib?
  https://github.com/microsoft/msquic/pull/2300 added *optional* support for high resolution timers, with *unconditional* linking of winmm.lib.
  AFAIU `timeGetDevCaps` etc. are not available for UWP apps,  
  https://learn.microsoft.com/en-us/windows/win32/api/timeapi/nf-timeapi-timegetdevcaps  
  (The use of this particular function isn't even guarded by the CMake option.)

File: ports/msh3/portfile.cmake [hunk 1]
```diff
@@ -1,20 +1,18 @@
...
 vcpkg_from_github(
     OUT_SOURCE_PATH SOURCE_PATH
     REPO nibanks/msh3
     REF v${VERSION}
-    SHA512 e6ba4e8f4ce5cd3f586d61739148bf75dfddbe70f399b2e498e7d416c8d730a5f8c2c38f0eabe687049bb7525df44f5f511515ec578bc3832989f73961cdda72
+    SHA512 dedd8be43e44b4bebbf601d76b1f3b0135501330ed128ca710de942ef7d9142a21f1c1eb9efecf57881e72d93d68c7c2c085bc35d402eac5eabc57e77773be6b
     HEAD_REF main
     PATCHES
         dependencies_fix.patch
+        width-exceeds-type.diff
```
Comment (dg0yt) [inline ports/msh3/portfile.cmake:11]: Fixes build error with gcc.

File: ports/msh3/portfile.cmake [hunk 2]
```diff
@@ -1,20 +1,18 @@
...
 )
 
-vcpkg_find_acquire_program(PKGCONFIG)
-
 vcpkg_cmake_configure(
     SOURCE_PATH "${SOURCE_PATH}"
-    OPTIONS
-        "-DPKG_CONFIG_EXECUTABLE=${PKGCONFIG}"
```
Comment (dg0yt) [inline ports/msh3/portfile.cmake:17]: Resolved by ls-qpack patches.

File: ports/msquic/portfile.cmake [hunk 3]
```diff
@@ -35,26 +36,33 @@ vcpkg_from_github(
     SHA512 1b26487fa79c8796d4b0d5e09f4fc9acb003d8e079189ec57a36ff03c9c2620829106fdbc4780e298872826f3a97f034d40e04d00a77ded97122874d13bfb145
     HEAD_REF main
 )
-
-file(REMOVE_RECURSE "${QUIC_SOURCE_PATH}/submodules/xdp-for-windows")
-file(RENAME "${XDP_WINDOWS}" "${QUIC_SOURCE_PATH}/submodules/xdp-for-windows")
+if(NOT EXISTS "${QUIC_SOURCE_PATH}/submodules/xdp-for-windows/published/external")
+    # headers only
+    file(REMOVE_RECURSE "${QUIC_SOURCE_PATH}/submodules/xdp-for-windows")
+    file(COPY "${XDP_WINDOWS}/published/external" DESTINATION "${QUIC_SOURCE_PATH}/submodules/xdp-for-windows/published")
```
Comment (dg0yt) [inline ports/msquic/portfile.cmake:42]: NB only using headers.

File: ports/msquic/uwp-link-libs.diff [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+index 4a573ae..6fb5887 100644
+--- a/src/platform/CMakeLists.txt
++++ b/src/platform/CMakeLists.txt
+@@ -60,7 +60,9 @@ if("${CX_PLATFORM}" STREQUAL "windows")
+         msquic_platform
+         PUBLIC
+         wbemuuid)
++    if(NOT QUIC_UWP_BUILD)
+     target_link_libraries(msquic_platform PUBLIC winmm)
++    endif()
```
Comment (dg0yt) [inline ports/msquic/uwp-link-libs.diff:24]: AFAIU API not supported on UWP.

File: ports/msquic/uwp-link-libs.diff [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+         if (${SYSTEM_PROCESSOR} STREQUAL "arm64")
+-            set(QUIC_OPENSSL_WIN_ARCH "VC-WIN64-ARM")
++            set(QUIC_OPENSSL_WIN_ARCH "VC-WIN64-ARM-UWP")
+         elseif (${SYSTEM_PROCESSOR} STREQUAL "arm")
+             set(QUIC_OPENSSL_WIN_ARCH "VC-WIN32-ARM")
+         elseif (${SYSTEM_PROCESSOR} STREQUAL "x86")
+             set(QUIC_OPENSSL_WIN_ARCH "VC-WIN32-ONECORE")
+         elseif (${SYSTEM_PROCESSOR} STREQUAL "x64" OR ${SYSTEM_PROCESSOR} STREQUAL "amd64")
+-            set(QUIC_OPENSSL_WIN_ARCH "VC-WIN64A-ONECORE")
++            set(QUIC_OPENSSL_WIN_ARCH "VC-WIN64A-UWP")
```
Comment (dg0yt) [inline ports/msquic/uwp-link-libs.diff:44]: AFAIU OneCore is not for UWP apps. Resolves VirtualAlloc linking errors.

**Comment by talregev** [comment]
@dg0yt What is missing that this PR will not be a draft?

**Review by MonicaLiu0311 — APPROVED** [review]

**Comment by talregev** [comment]
@MonicaLiu0311 Did you also test this port?
If yes, can you add a review label?

**Comment by MonicaLiu0311** [comment]
```
PS E:\msh3> ./vcpkg install vcpkg-ci-msh3
Computing installation plan...
The following packages will be built and installed:
  * msh3:x64-windows@0.7.0
    vcpkg-ci-msh3:x64-windows@ci
Additional packages (*) will be modified to complete this operation.
Detecting compiler hash for triplet x64-windows...
Compiler found: C:/Program Files/Microsoft Visual Studio/2022/Enterprise/VC/Tools/MSVC/14.41.34120/bin/Hostx64/x64/cl.exe
Installing 1/2 msh3:x64-windows@0.7.0...
Building msh3:x64-windows@0.7.0...
-- Using cached nibanks-msh3-v0.7.0.tar.gz.
-- Cleaning sources at E:/msh3/buildtrees/msh3/src/v0.7.0-5ad3078e60.clean. Use --editable to skip cleaning for the packages you specify.
-- Extracting source E:/msh3/downloads/nibanks-msh3-v0.7.0.tar.gz
-- Applying patch dependencies_fix.patch
-- Applying patch width-exceeds-type.diff
-- Using source at E:/msh3/buildtrees/msh3/src/v0.7.0-5ad3078e60.clean
-- Found external ninja('1.11.0').
-- Configuring x64-windows
-- Building x64-windows-dbg
-- Building x64-windows-rel
-- Fixing pkgconfig file: E:/msh3/packages/msh3_x64-windows/lib/pkgconfig/libmsh3.pc
-- Using cached msys2-mingw-w64-x86_64-pkgconf-1~2.3.0-1-any.pkg.tar.zst.
-- Using cached msys2-msys2-runtime-3.5.4-2-x86_64.pkg.tar.zst.
-- Using msys root at E:/msh3/downloads/tools/msys2/21caed2f81ec917b
-- Fixing pkgconfig file: E:/msh3/packages/msh3_x64-windows/debug/lib/pkgconfig/libmsh3.pc
-- Installing: E:/msh3/packages/msh3_x64-windows/share/msh3/copyright
-- Performing post-build validation
Stored binaries in 1 destinations in 199 ms.
Elapsed time to handle msh3:x64-windows: 6.7 s
msh3:x64-windows package ABI: 118188eea1376f2ec72d1f6cf3e731eb3b165d3a4934f54d024d15df2a7ff575
Installing 2/2 vcpkg-ci-msh3:x64-windows@ci...
Building vcpkg-ci-msh3:x64-windows@ci...
-- Using cached msys2-mingw-w64-x86_64-pkgconf-1~2.3.0-1-any.pkg.tar.zst.
-- Using cached msys2-msys2-runtime-3.5.4-2-x86_64.pkg.tar.zst.
-- Using msys root at E:/msh3/downloads/tools/msys2/21caed2f81ec917b
-- Found external ninja('1.11.0').
-- Configuring x64-windows
-- Building x64-windows-dbg
-- Building x64-windows-rel
-- Skipping post-build validation due to VCPKG_POLICY_EMPTY_PACKAGE
Stored binaries in 1 destinations in 29.3 ms.
Elapsed time to handle vcpkg-ci-msh3:x64-windows: 2.5 s
vcpkg-ci-msh3:x64-windows package ABI: 1d55859fe8960fded38db8b304184a7d4621df1e82e554b430628fcd8856b837
Total install time: 9.1 s
```

**Comment by talregev** [comment]
@dg0yt Can you update and create PRs in msquic upstream with your changes?

**Comment by talregev** [comment]
@MonicaLiu0311 Can you add depends:upstream-changes label? It needed to create PRs on msquic upstream first.

**Comment by dg0yt** [comment]
> @dg0yt Can you update and create PRs in msquic upstream with your changes?

> @MonicaLiu0311 Can you add depends:upstream-changes label? It needed to create PRs on msquic upstream first.

@talregev I kindly ask to stop that kind of advice.

This PR is a pure improvement over what vcpkg has now. This PR is a finished work, offering an update plus an automated test for CMake usage.

From the start of this work, I submitted PRs for ls-qpack and msh3. Upstreaming msquic changes is pending because this repo needs different modification than the vcpkg port. The HEAD of msquic already moved away from the release which is used in this PR. Improving/fixing that will be more or less an independent effort, not something offering a clear indication of upstream's acceptance of the patches here.

And now remember that I don't have a MSVC toolchain, only vcpkg CI. To use upstream's resource respectfully, I have to consider cases which are not covered in my local builds.

**Comment by nibanks** [comment]
@dg0yt I totally understand your perspective here. I wouldn't block this PR for those up level changes. But I do ask you consider the security and quality angle here involving patches being applied to projects without most of the people who own those projects actually reviewing those patches. A different group reviews these vcpkg PRs than those who review MsQuic. We have different automation running in each also. By only applying these changes on top via patches, we have less confidence that things are all correct, which can create security and quality issues. So, we should strive for getting changes in the upstream projects instead of via patch files. But like I said, it shouldn't be an absolute requirement. Flexibility allows for forward progress at a faster pace.

**Comment by dg0yt** [comment]
@nibanks There is no disagreement about that.
OTOH reviewing the past work gave me little confidence that ports and upstreams are taking the right direction for CMake config export and for UWP, despite all infrastructure. That's why this PR started test-first: Adding a central piece of infrastructure before making changes.

BTW the questions in https://github.com/microsoft/vcpkg/pull/42788#issuecomment-2558072453 were meant to save upstream review iterations by avoiding to start with the wrong guess.

**Pull request merged**